### ISSUE TITLE
AS v3.8.1 in Docker fails to deploy to targetHost

### ISSUE THREAD

**Issue by ajsq** [F5Networks/f5-appsvcs-extension#64]
##### ISSUE TYPE
 - Bug Report

##### AS3 BUILD/ VERSION
```
Docker container v3.8.1
```

##### BIGIP VERSION

```
Sys::Version
Main Package
  Product     BIG-IP
  Version     14.1.0.1
  Build       0.0.7
  Edition     Point Release 1
  Date        Tue Jan 15 19:49:59 PST 2019
```

##### SUMMARY

```
AS3 v3.8.1 fails to deploy to a target system.
```

##### STEPS TO REPRODUCE

Attempt to deploy to target (which does not have AS3 installed) via AS3 3.8.1 in container. AS3 returns error above.

Example deployment in [gist](https://gist.github.com/ajsq/e8d254af9ed2f3d9bf406560305a6d3b)

Reproducible with AS3 3.8.1 installed on BIG-IP as well, but I'm not clear on whether using the targetHost etc options are supported when using AS3 installed on BIG-IP.

##### EXPECTED RESULTS

Deploy one virtual server to target BIG-IP, similar to locally installed AS3.

##### ACTUAL RESULTS

response from AS3:
```json
{
   "code" : 500,
   "declarationFullId" : "",
   "message" : "Status Code 401 POST /mgmt/shared/file-transfer/uploads/f5-appsvcs-discovery-1.0.0-7.noarch.rpm HTTP/1.1\r"
}
```

log excerpt (from container /var/log/restnoded/restnoded.log) :
```
Wed, 13 Feb 2019 01:24:27 GMT - fine: [appsvcs] {"message":"got token for admin@10.3.67.176","level":"debug"}
Wed, 13 Feb 2019 01:24:27 GMT - info: [appsvcs] {"message":"modules provisioned: ltm","level":"info"}
Wed, 13 Feb 2019 01:24:27 GMT - info: [appsvcs] {"message":"target device is BIG-IP version 14.1.0.1.0.0.7","level":"info"}
Wed, 13 Feb 2019 01:24:27 GMT - info: [appsvcs] {"message":"Installing cloud-libs and service discovery worker","level":"info"}
Wed, 13 Feb 2019 01:24:28 GMT - fine: [appsvcs] {"message":"Installing f5-cloud-libs-consul.tar.gz to remote device","level":"debug"}
Wed, 13 Feb 2019 01:24:28 GMT - fine: [appsvcs] {"message":"Installing f5-cloud-libs.tar.gz to remote device","level":"debug"}
Wed, 13 Feb 2019 01:24:29 GMT - fine: [appsvcs] {"message":"Installing f5-cloud-libs-aws.tar.gz to remote device","level":"debug"}
Wed, 13 Feb 2019 01:24:29 GMT - fine: [appsvcs] {"message":"Installing f5-cloud-libs-gce.tar.gz to remote device","level":"debug"}
Wed, 13 Feb 2019 01:24:29 GMT - fine: [appsvcs] {"message":"Installing f5-cloud-libs-azure.tar.gz to remote device","level":"debug"}
Wed, 13 Feb 2019 01:24:34 GMT - info: [appsvcs] {"message":"Error: Status Code 401 POST /mgmt/shared/file-transfer/uploads/f5-appsvcs-discovery-1.0.0-7.noarch.rpm HTTP/1.1\r","stack":["Error: Status Code 401 POST /mgmt/shared/file-transfer/uploads/f5-apps
vcs-discovery-1.0.0-7.noarch.rpm HTTP/1.1\r","checkForHttpError (/var/config/rest/iapps/f5-appsvcs/nodejs/utils/iappUtil.js:55:21)","IncomingMessage.res.on (/var/config/rest/iapps/f5-appsvcs/nodejs/utils/iappUtil.js:74:31)","emitNone (events.js:111:20)","
IncomingMessage.emit (events.js:208:7)","endReadableNT (_stream_readable.js:1064:12)","_combinedTickCallback (internal/process/next_tick.js:138:11)","process._tickCallback (internal/process/next_tick.js:180:9)"],"level":"info"}
```

**Comment by ajsq** [comment]
Test as suggested in Slack: same issue occurs if `targetPassphrase`/`targetUsername` is replaced with `targetTokens`

**Comment by jikui** [comment]
I also run into the same issue when I tried to deploy the VIP and pool memebers.
1) The response from the AS3
{
    "code": 500,
    "declarationFullId": "",
    "message": "Status Code 401 POST /mgmt/shared/file-transfer/uploads/f5-appsvcs-discovery-1.0.0-7.noarch.rpm HTTP/1.1\r"
}
2) The syntax scheme
{
    "class": "AS3",
    "action": "deploy",
    "targetHost": "10.128.0.149",
    "targetUsername": "admin",
    "targetPassphrase": "admin",
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.8.0",
        "id": "Service_HTTP",
        "Sample_container": {
            "class": "Tenant",
            "Application": {
                "class": "Application",
                "template": "Service_Generic",
                "ItemTest": {
                    "class": "Service_HTTP",
                    "policyEndpoint": "endpointPolicy",
                    "virtualAddresses": ["10.128.20.222"],
                    "pool": "web_pool"
                },
                "web_pool": {
                    "class": "Pool",
                    "monitors": [
                        "http"
                    ],
                    "members": [{
                        "servicePort": 80,
                        "serverAddresses": [
                            "10.128.10.145"
                          ]
                    }]
                },
                 "endpointPolicy": {
                    "class": "Endpoint_Policy",
                    "rules": [
                        {
                            "name": "rule1",
                            "conditions": [{
                                "type": "httpUri",
                                "path": {
                                	"operand": "contains",
                                    "values": ["/test1/"]
                                }
                            }
                            ],
                            "actions": [{
                                "type": "waf",
                                "policy": {
                                    "use": "asm_demo"
                                }
                            }
                            ]
                        },
                        {
                            "name": "rule2",
                            "conditions": [{
                                "type": "httpUri",
                                "path": {
                                	"operand": "contains",
                                    "values": ["/test2/"]
                                }
                            }
                            ],
                            "actions": [{
                                "type": "waf",
                                "policy": {
                                    "use": "multi_policy"
                                }
                            }
                            ]
                        },
                       {
                            "name": "default",
                            "conditions": [],
                            "actions": [{
                            	"type": "waf"
                               }
                            ]
                        }
                    ]
                },
                "asm_demo": {
                    "class": "WAF_Policy",
                    "url": "http://10.145.106.100/asm-policy/Demo_asm_demo.xml"
                },
                "multi_policy": {
                    "class": "WAF_Policy",
                    "url": "http://10.145.106.100/asm-policy/Demo_multi_policy.xml"
                }
                
                }
            
            }
        }
    
}

**Comment by tkam8** [comment]
Facing same issue here too. 
However, this does not reproduce on v3.5.1 (long term support version), so this may be a regression of some sort.

**Comment by jmcalalang** [comment]
Isn’t 14.1 not supported? Did I miss that announcement? I wonder if your hitting some kind of object incompatibility

Cheers,

Jon Calalang

________________________________

**Comment by tkam8** [comment]
Adding that I repro'd on BIG-IP version v13.1.0.2
# tmsh show sys software

--------------------------------------------------
Sys::Software Status
Volume  Product   Version  Build  Active    Status
--------------------------------------------------
HD1.1    BIG-IP  13.1.0.2  0.0.6     yes  complete

**Comment by ajsq** [comment]
> Isn’t 14.1 not supported? Did I miss that announcement?

The [Prerequisites and Requirements][] page says only 12.1.x or later. 

AS3 v3.8.1 works if installed directly on a v14.1 BIG-IP and deploying to the local device, and this issue does not occur with AS3 v3.7.0.

[Prerequisites and Requirements]: https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/userguide/prereqs.html

**Comment by dstokesf5** [comment]
This looks similar to an issue that should be fixed in 3.8.1. Tracking this internally now to look further into the issue.

**Comment by jikui** [comment]
> > Isn’t 14.1 not supported? Did I miss that announcement?
> 
> The [Prerequisites and Requirements](https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/userguide/prereqs.html) page says only 12.1.x or later.
> 
> AS3 v3.8.1 works if installed directly on a v14.1 BIG-IP and deploying to the local device, and this issue does not occur with AS3 v3.7.0.

yes, even AS3.8.0 works well if it is installed in bigip. The issue occurs if AS3 works as a container.

Thanks.

**Comment by kaustriaf5** [comment]
Tracked internally as 901, fixed slated for 3.10

**Comment by kaustriaf5** [comment]
Hi @ajsq , @jikui ,
The fix has been released in [3.10](https://github.com/F5Networks/f5-appsvcs-extension/releases/tag/v3.10.0). 
Closing this issue.
